pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -164,6 +164,15 @@ func (t ResolvedPipelineRunTask) IsCancelled() bool {
 	return c != nil && c.IsFalse() && c.Reason == v1beta1.TaskRunReasonCancelled.String()
 }
 
+// IsScheduled returns true when the PipelineRunTask itself has a TaskRun
+// or Run associated.
+func (t ResolvedPipelineRunTask) IsScheduled() bool {
+	if t.IsCustomTask() {
+		return t.Run != nil
+	}
+	return t.TaskRun != nil
```

Comment:
@afrittoli @jerop @vdemeester my understanding was this is not possible 🙃  A `taskRun` or `run` is associated must have a `succeeded` condition with `unknown` i.e. `running` status.

We rely on the same check for the non-final tasks:

https://github.com/tektoncd/pipeline/blob/4ad8f33f8f30ccd4ffc22d57d16dd87420721094/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L207

Reply:
I'm still not sure about the root cause (my current assumption is a timing issue), but I think it makes sense to check if `Run` exists only in the `IsStarted()` method like below.

```go
func (t ResolvedPipelineRunTask) IsStarted() bool {
	if t.IsCustomTask() {
		return t.Run != nil

	}
	return t.TaskRun != nil
}
```

Whether it has a `Succeeded-type condition` or not, I think we can assume the task is started since the reconciler found that it has associated `Run`.

What do you think?

Reply:
@devholic makes sense to go with that option 👍🏾 

cc @tektoncd/core-maintainers

Reply:
Yeah, the `TaskRun` condition is set by the `TaskRun` reconciler on the first reconciliation loop. 
In the `TaskRun` controller, at least within the `ReconcileKind` function, the condition will always be set as initialising the condition is the first thing we do in `ReconcileKind`.
In the `PipelineRun` controller instead there is no guarantee. It may not be very likely to catch a `TaskRun`  with no condition set, but it is certainly possible and I guess it may happen more under high load.